Introduce Redfish http protocol and library

nicklela opened the pull request:
Introduce Redfish HTTP protocol to improve Redfish performance
and communication stability between BIOS and Redfish service.

Signed-off-by: Nickle Wang [nicklew@nvidia.com](mailto:nicklew@nvidia.com)
Co-authored-by: : Igor Kulchytskyy [igork@ami.com](mailto:igork@ami.com)
Cc: Abner Chang [abner.chang@amd.com](mailto:abner.chang@amd.com)

nicklela committed:
RedfishPkg: introduce RedfishHttpLib

RedfishHttpLib is a wrapper library for Redfish feature drivers to
call Redfish HTTP Protocol easily.

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>
Cc: Nick Ramirez <nramirez@nvidia.com>

nicklela committed:
RedfishPkg/RedfishLib: include RedfishServiceData.h

Redfish common structures are moved to RedfishServiceData.h. Remove
them from RedfishLib.h

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>

nicklela committed:
RedfishPkg: introduce Redfish HTTP protocol

Introduce Redfish HTTP protocol to improve Redfish performance
and communication stability between BIOS and Redfish service.
- Feature drivers often query same Redfish resource multiple
times for different purpose. Implement HTTP cache mechanism to
improve HTTP GET performance. "UseCache" parameter is provided
if application likes to send HTTP GET request to Redfish service
without using cache data.
- This driver will retire stale cache data automatically when
application modify Redfish resource at service side.
- PCD PcdHttpCacheDisabled is used to disable cache mechanism in
this driver for debugging purpose.
- PCD PcdRedfishServiceContentEncoding is used to enable content
encoding while sending data to Redfish service.
- Redfish HTTP protocol also implement retry mechanism to retry
HTTP request when BIOS receive unexpected response from Redfish service.
This function helps BIOS Redfish to finish its job as much as possible.
- PCDs are defined to control how many times BIOS will retry the
request and how many time BIOS will wait between retries.

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Co-authored-by: Igor Kulchytskyy <igork@ami.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>

nicklela committed:
RedfishPkg: implement Redfish HTTP protocol

implement Redfish HTTP protocol driver.

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Co-authored-by: Igor Kulchytskyy <igork@ami.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>

nicklela committed:
RedfishPkg/RedfishDebugLib: use RedfishHttpLib

Remove RedfishLib and use RedfishHttpLib for debug printing
Redfish response data.

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>

changab commented on RedfishPkg/Include/Library/RedfishHttpLib.h:
```diff
@@ -0,0 +1,328 @@
+  after this function returns successfully.
+
+  @param[in]  This            Pointer to EDKII_REDFISH_HTTP_PROTOCOL instance.
+  @param[in]  RedfishService  Pointer to Redfish service to be released.
```
[in] for This is not needed.

changab commented on RedfishPkg/Include/Library/RedfishHttpLib.h:
```diff
@@ -0,0 +1,328 @@
+  is a reference to the JSON value in RedfishPayload. Any modification to returned
+  JSON value will change JSON value in RedfishPayload.
+
+  @param[in]  This            Pointer to EDKII_REDFISH_HTTP_PROTOCOL instance.
```
[in] for This is not needed.

changab commented on RedfishPkg/Include/Library/RedfishHttpLib.h:
```diff
@@ -0,0 +1,328 @@
+  supported. It's caller's responsibility to Response by calling
+  RedfishHttpFreeResponse ().
+
+  @param[in]  Service       Redfish service instance to make query.
```
Is the comment "Redfish service instance to make query." accurate? Or "Redfish service instance to perform HTTP GET"?

changab commented on RedfishPkg/Include/Library/RedfishHttpLib.h:
```diff
@@ -0,0 +1,328 @@
+  Perform HTTP PATCH to send redfish resource to given resource URI.
+  It's caller's responsibility to free Response by calling RedfishHttpFreeResponse ().
+
+  @param[in]  Service       Redfish service instance to make query.
```
Is the comment "Redfish service instance to make query." accurate? Or "Redfish service instance to perform HTTP PATCH"?

changab commented on RedfishPkg/Include/Library/RedfishHttpLib.h:
```diff
@@ -0,0 +1,328 @@
+  Perform HTTP PATCH to send redfish resource to given resource URI.
+  It's caller's responsibility to free Response by calling RedfishHttpFreeResponse ().
+
+  @param[in]  Service       Redfish service instance to make query.
```
Is the comment "Redfish service instance to make query." accurated? Or "Redfish service instance to perform HTTP PATCH"?
Please check the same occurrence in the rest of files.

changab commented on RedfishPkg/Library/RedfishHttpLib/RedfishHttpLib.c:
```diff
@@ -0,0 +1,585 @@
+    return;
+  }
+
+  gBS->CloseEvent (Event);
```
Would you like to move this to the beginning of this function? Also remove the one when if (mRedfishHttpProtocol != NULL)?

nicklela replied:
We only close the event when we can locate Redfish HTTP protocol, right?

changab replied:
But we should also close the event when we locate HTTP protocol successfully?

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+  @param[in]  ContentSize   Size of the Content to be send to Redfish service.
+                            This is optional. When ContentSize is 0, ContentSize
+                            is the size of Content.
+  @param[in]  ContentType   Type of the Content to be send to Redfish service.
```
Could we list the supported ContentType in the function header? This would be more frindly to the funciton users.

nicklela replied:
We don't do the check to Content Type. Since this is optional, I can add default content type that we will use here.

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+  @param[in]  ContentSize   Size of the Content to be send to Redfish service.
+                            This is optional. When ContentSize is 0, ContentSize
+                            is the size of Content.
+  @param[in]  ContentType   Type of the Content to be send to Redfish service.
```
Could we list the supported ContentType in the function header? This would be more friendly to the function users.
Same as the below occurrences.

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+
+  Request.Content       = Content;
+  Request.ContentLength = ContentSize;
+  Request.ContentType   = ContentType;
```
Do we have to check check if the Content type is supported or not?

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+
+  Private = (REDFISH_HTTP_CACHE_PRIVATE *)Context;
+  if (Private->Signature != REDFISH_HTTP_DRIVER_SIGNATURE) {
+    return;
```
Could we add a debug message for this error scenario? As this may considered a fatal error.

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+                  (VOID **)&Private->CredentialProtocol
+                  );
+  if (EFI_ERROR (Status)) {
+    return;
```
Could we add a debug message for this error scenario? As this may considered a fatal error.

nicklela replied:
There is chance to locate protocol with failure.  EfiCreateProtocolNotifyEvent() will at lease call this callback function nonce.  And this is not a failure.

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+    return;
+  }
+
+  gBS->CloseEvent (Event);
```
We can move this to the beginning of funciton.

nicklela replied:
No, I think we only close event when we can successfully locate protocol.

changab replied:
My thought is when we get the notification of Redfish HTTP protocol installation. We should be able to locate the protocol. If not, then there must be something wrong. Suppose we won't get another notification of protocol installation as there is only one instance of Redfish HTTP protocol, right?

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpDxe.c:
```diff
@@ -0,0 +1,1336 @@
+                                            );
+  if (mRedfishHttpCachePrivate->NotifyEvent == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a: failed to create protocol notification for gEdkIIRedfishCredentialProtocolGuid\n", __func__));
+    ASSERT (FALSE);
```
we should also return EFI error in the case ASSERT is implicitly disabled.  Also free mRedfishHttpCachePrivate memory.

changab commented on RedfishPkg/RedfishHttpDxe/RedfishHttpOp.h:
```diff
@@ -0,0 +1,75 @@
+/** @file
+  Definitions of RedfishHttpOp
```
What the RedfishHttpOp refers? We may want to give it an understandable naming.

nicklela replied:
That is Redfish Http Operation functions. Does RedfishHttpOperation.h sound good?

changab replied:
Yes,  RedfishHttpOperation.h is good.

changab reviewed (commented):
I had left comments in the files.

nicklela committed:
RedfishPkg/RedfishCrtLib: fix unresolved external symbol issue

-Fix below compiler error reported in edk2 CI.
ERROR - Linker #2001 from JsonLib.lib(load.obj) : unresolved external
symbol __ftol2
-The file MathFtol.c is copied from IntrinsicLib in CryptoPkg.
-Add MathFtol.c to EccCheck IgnoreFiles.

Signed-off-by: Nickle Wang <nicklew@nvidia.com>
Cc: Abner Chang <abner.chang@amd.com>
Cc: Igor Kulchytskyy <igork@ami.com>

changab reviewed (commented):
I gave my responses to the commnet.
